Add friendly id to user and tip #8

MrBeardface opened the pull request:
Things to evaluate:
1. Still need to check whether duplicate copies are being made of username and tips. Believe that this is solved with the #should_generate_new_friendly_id? in the User and Tip model.
2. In production, will all the current Users and Tips be updated with Friendly Id? Or do we have to create a Rake Task?

MrBeardface committed:
Add Friendly Id for Users

- Add Friendly Id gem to Gemfile
- Generate migration on User table for slug
- Update #set_user method on User controller

MrBeardface committed:
Add method to User model

- Add #should_generate_new_friendly_id? to check if user is new or slug
  is blank.

MrBeardface committed:
Add Friendly Id to Tips

- Add friendly to Tip controller #set_tip
- Create custom slug for tips
- Generate migration on Tip table for slug

ohlincik commented on Gemfile:
```diff
@@ -44,6 +44,8 @@ gem 'sitemap_generator'
 # Makes sending HTML emails a little less painful by inlining stylesheets and rewriting relative URLs
 gem 'roadie-rails', '~> 1.0.4'
 
+gem 'friendly_id', '~> 5.1.0'
```
It is a good idea to provide a comment what this gem is for and sometimes even provide a link to the repo on GitHub or documentation. See all the other gems above that.

ohlincik commented:
I don't think we need to worry about the username duplicates. These cannot be changed by the user and we are getting them straight from GitHub so they are guaranteed unique.

It doesn't look like the `should_generate_new_friendly_id?` is handling the uniqueness. It seems like it handles the creation of new slug in the case the record is new of if the slug is empty.

I believe that all existing users and tips will require to have the the slugs generated upon deployment. This can be handled by a rake task or possibly packaging it into a migration (not sure if that is a good idea).

MrBeardface commented:
When I go into the Rails console and try to create a new record with the same Tip Title as another Tip, the creations fails. If I take out should_generate_new_friendly_id? and try again, it still fails.

If I try to create a New Tip in development with the same Tip Title as another Tip, it generates it with a UUID ex) tip-time-by-mrbeardface-b4ba2d4e-6c6d-4306-bd9c-50972ae1bdbd.

Do we want this or no?

MrBeardface committed:
Update Friendly_id functionality

- Add description for Friendly_id gem in Gemfile
- Add #check_tip_slugs to find out if #New Tip is actually a duplicate.
- Add alert so users know why their Tip didn't save

MrBeardface commented:
Things that still need to be done:

1. Is this setup SEO friendly?
2. Rake task needs to be setup for Users and Tips in production.

ohlincik committed:
Modify user and tip migrations

- Create index for User slugs
- Generate the slugs for existing users
- Create index for Tip slugs
- Generate the slugs for existing tips

ohlincik committed:
Refactor friendly_id functionality

- Simplify the title uniqueness validation for Tips
- Add restriction for generating new slugs for Tips before 1 day passes
  since their creation

ohlincik committed:
Fix small bug in friendly_id for Tips

The slug was not being generated for brand new tips

ohlincik committed:
Fix another small bug in friendly_id for Tips

Apparently the condition for generating new slugs was reversed.

ohlincik commented:
I think this should be in good shape. Please test this for yourself and if all is good, you can merge.
